### PR TITLE
Binary Search Tree

### PR DESCRIPTION
**Pull request by tgrander**
Binary Search Tree data structure

### CODE REVIEW

File: lib/data-structures/BinarySearchTree.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    if (value > nodeValue) {
+      return this.search(value, node.getRight());
+    }
+    return this.search(value, node.getLeft());
+  }
+
+  /**
+   * Recrusively get the minimum value in the tree.
```
Comment (yangshun) [inline lib/data-structures/BinarySearchTree.js]: Typo - `Recrusively`

File: lib/data-structures/BinarySearchTree.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
+import BinaryTreeNode from './BinaryTreeNode';
+
+class BinarySearchTree {
+  /**
+   * Initialize the Binary Search Tree (BST)
+   * @param {*} value The value
+   */
+  constructor(value) {
+    if (value) {
+      this._root = new BinaryTreeNode(value);
```
Comment (yangshun) [inline lib/data-structures/BinarySearchTree.js]: Prefer early return here - can remove the `else`

File: lib/data-structures/BinarySearchTree.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+  /**
+   * Recursively get the maximum value in the tree.
+   * @return {*} The maximum value.
+   */
+  getMaximum(node = this._root) {
+    if (!node) return null;
+
+    const right = node.getRight();
+    if (!right) return node;
```
Comment (yangshun) [inline lib/data-structures/BinarySearchTree.js]: Shouldn't this return `node.getValue()`?

File: lib/data-structures/BinarySearchTree.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+      return node.setRight(new BinaryTreeNode(value));
+    }
+    return this.insert(value, right);
+  }
+
+  /**
+   * Recursively search for a value in the BST
+   * @param {*} value The value to search for.
+   * @param {*} node The current node.
+   * @return {*} Either the value or null if value is not found
```
Comment (yangshun) [inline lib/data-structures/BinarySearchTree.js]: I think search should return `true`/`false` to indicate existence.

**Review by yangshun — CHANGES_REQUESTED** [review]
Thanks! This looks very good in general. Just a couple of small changes before merging 👍

**Comment by tgrander** [comment]
All comments responded to!

**Comment by tgrander** [comment]
Also updated tests to reflect changes

File: lib/data-structures/BinarySearchTree.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
```
Comment (yangshun) [inline lib/data-structures/BinarySearchTree.js]: The doc's return type needs to be updated.

File: lib/data-structures/BinarySearchTree.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    if (value) {
+      this._root = new BinaryTreeNode(value);
+    }
+  }
+
+  /**
+   * Recursively insert a new value in the BST.
+   * @param {*} value The value being inserted
+   * @param {BinaryTreeNode} node The current node. Param is not required.
+   * @return {*}
```
Comment (yangshun) [inline lib/data-structures/BinarySearchTree.js]: Return `undefined`

**Review by yangshun — CHANGES_REQUESTED** [review]
Just some final nits, thanks a lot 😄

**Review by yangshun — APPROVED** [review]

**Comment by yangshun** [comment]
Many thanks! 🎉

**Pull request merged**